### PR TITLE
Add broadcast support for softmax_cross_entropy_with_logits

### LINKED ISSUE #11534: [FeatureRequest] make categorical_x_entropy broadcastable
For example:

```
>>> x = tf.placeholder('float32', [4,5,6])
>>> y = tf.placeholder('float32', [1,5,6])
>>> z = tf.nn.softmax_cross_entropy_with_logits(labels=x, logits=y)
Traceback (most recent call last):
  File "/home/khaotik/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/common_shapes.py", line 671, in _call_cpp_shape_fn_impl
    input_tensors_as_shapes, status)
  File "/home/khaotik/anaconda3/lib/python3.6/contextlib.py", line 89, in __exit__
    next(self.gen)
  File "/home/khaotik/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/errors_impl.py", line 466, in raise_exception_on_not_ok_status
    pywrap_tensorflow.TF_GetCode(status))
tensorflow.python.framework.errors_impl.InvalidArgumentError: Dimension 0 in both shapes must be equal, but are 5 and 20 for 'SoftmaxCrossEntropyWithLogits' (op: 'SoftmaxCrossEntropyWithLogits') with input shapes: [5,6], [20,6].

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/khaotik/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/nn_ops.py", line 1594, in softmax_cross_entropy_with_logits
    precise_logits, labels, name=name)
  File "/home/khaotik/anaconda3/lib/python3.6/site-packages/tensorflow/python/ops/gen_nn_ops.py", line 2380, in _softmax_cross_entropy_with_logits
    features=features, labels=labels, name=name)
  File "/home/khaotik/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/op_def_library.py", line 767, in apply_op
    op_def=op_def)
  File "/home/khaotik/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 2508, in create_op
    set_shapes_for_outputs(ret)
  File "/home/khaotik/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1873, in set_shapes_for_outputs
    shapes = shape_func(op)
  File "/home/khaotik/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/ops.py", line 1823, in call_with_requiring
    return call_cpp_shape_fn(op, require_shape_fn=True)
  File "/home/khaotik/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/common_shapes.py", line 610, in call_cpp_shape_fn
    debug_python_shape_fn, require_shape_fn)
  File "/home/khaotik/anaconda3/lib/python3.6/site-packages/tensorflow/python/framework/common_shapes.py", line 676, in _call_cpp_shape_fn_impl
    raise ValueError(err.message)
ValueError: Dimension 0 in both shapes must be equal, but are 5 and 20 for 'SoftmaxCrossEntropyWithLogits' (op: 'SoftmaxCrossEntropyWithLogits') with input shapes: [5,6], [20,6].
```

But, intuitively, I'd like to broadcast `y` into shape `[4,5,6]` in the above code.

It's not too hard to get a workaround such as using `tf.tile`. However it would consume more precious GPU memory.

### PR DESCRIPTION
**Pull request by yongtang**
This fix tries to address the issue raised in #11534 where there was no broadcast support for SoftmaxCrossEntropyWithLogits. This fix adds the broadcast support for SoftmaxCrossEntropyWithLogits, and adds test cases for it.

This fix fixes #11534.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: tensorflow/core/kernels/xent_op.h [hunk 1]
```diff
@@ -84,10 +93,12 @@ struct XentEigenImpl {
 #endif
 
     // max_logits along classes.
-    scratch.reshape(batch_only).device(d) = logits.maximum(along_class);
+    scratch.reshape(batch_only).device(d) =
```
Comment (rmlarsen) [inline tensorflow/core/kernels/xent_op.h]: Could you please verify that the presence of the extra broadcast expression does not affect performance in the trivial case? If it does, perhaps we should have a separate code path for that. This is a pretty performance critical op for many models.

File: tensorflow/core/ops/nn_ops.cc [hunk 2]
```diff
@@ -1062,12 +1062,21 @@ REGISTER_OP("SoftmaxCrossEntropyWithLogits")
     .Attr("T: {half, bfloat16, float, double}")
     .SetShapeFn([](InferenceContext* c) {
       ShapeHandle input;
-      TF_RETURN_IF_ERROR(c->WithRank(c->input(0), 2, &input));
-      TF_RETURN_IF_ERROR(c->Merge(input, c->input(1), &input));
+      if (c->WithRank(c->input(0), 2, &input) == Status::OK() && c->Merge(input, c->input(1), &input) == Status::OK()) {
```
Comment (rmlarsen) [inline tensorflow/core/ops/nn_ops.cc]: Please keep lines below 80 characters. Use clangformat i in doubt.

File: tensorflow/core/ops/nn_ops_test.cc [hunk 3]
```diff
@@ -410,10 +410,8 @@ TEST(NNOpsTest, SoftmaxCrossEntropyWithLogits_ShapeFn) {
   INFER_OK(op, "[1,?];[?,2]", "[d0_0];[d0_0,d0_1|d1_1]");
   INFER_OK(op, "[?,2];[1,2]", "[d1_0];in1");
 
-  INFER_ERROR("Dimension 0 in both shapes must be equal, but are 1 and 2", op,
-              "[1,?];[2,?]");
-  INFER_ERROR("Shape must be rank 2 but is rank 3", op, "[1,2,3];?");
-  INFER_ERROR("Shapes must be equal rank, but are 2 and 3", op, "?;[1,2,3]");
+  INFER_ERROR("Shape must be broadcasted with rank 2", op, "[1,2,3];?");
+  INFER_ERROR("Shape must be broadcasted with rank 2", op, "?;[1,2,3]");
```
Comment (rmlarsen) [inline tensorflow/core/ops/nn_ops_test.cc]: Could you add cases with successful broadcast here?

**Review by rmlarsen — CHANGES_REQUESTED** [review]

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rmlarsen** [comment]
@yongtang any followup on this? (also, please rebase)

File: tensorflow/core/ops/nn_ops_test.cc (hunk 3 above)
Comment (yongtang) [reply tensorflow/core/ops/nn_ops_test.cc]: @rmlarsen Thanks. Additional test cases have been added.

File: tensorflow/core/ops/nn_ops.cc (hunk 2 above)
Comment (yongtang) [reply tensorflow/core/ops/nn_ops.cc]: Thanks. Done.

File: tensorflow/core/kernels/xent_op.h (hunk 1 above)
Comment (yongtang) [reply tensorflow/core/kernels/xent_op.h]: Thanks @rmlarsen. I added zero dimensions and single classes for benchmark. No significant performance changes so far:

Before:
```
Benchmark: zero_dimension_m_128_n_100000_k_100000_use_gpu_False 	 wall_time: 0.000133 s 	 Throughput: 96.2 GB/s

Benchmark: single_class_m_128_n_100000_k_100000_use_gpu_False 	 wall_time: 0.000131 s 	 Throughput: 97.6 GB/s

```

After:
```
Benchmark: zero_dimension_m_128_n_100000_k_100000_use_gpu_False 	 wall_time: 0.000135 s 	 Throughput: 94.7 GB/s

Benchmark: single_class_m_128_n_100000_k_100000_use_gpu_False 	 wall_time: 0.000133 s 	 Throughput: 96.4 GB/s
```

**Review by rmlarsen — APPROVED** [review]

**Comment by rmlarsen** [comment]
@yongtang Thanks for the contribution.

**Comment by rmlarsen** [comment]
@yongtang This PR is showing a significant performance regression in our internal monitoring. Can you please take a look? 

Compare the benchmark before/after by running:

bazel test tensorflow/core/kernels:xent_op_test --test_arg=--benchmarks=all

Thanks.

**Comment by annarev** [comment]
@yongtang To add to @rmlarsen's command, you can pass --test_output=streamed to print out benchmark outputs.

I just tried to run the benchmarks at and before the commit in this PR and can see the regression. For e.g. BM_Xent_16_10000_gpu benchmark prints:
570.9M items/s without this change
373.9M items/s with this change

(I was running with 4 Titan X gpus, so numbers are probably different with other configs)

**Comment by rmlarsen** [comment]
@yongtang It's OK. I have a fix for it. It will be pushed to open source sometime next week.

**Comment by rmlarsen** [comment]
@yongtang To be clear: I am improving Eigen's broadcasting to handle the trivial case better. We have to upstream this to Eigen before it will appear in TensorFlow, but expect it to land sometime next week.

**Comment by yongtang** [comment]
@rmlarsen @annarev Thanks a lot for the help and sorry for the trouble the PR was causing. Please let me know if there are anything else I could help.

**Pull request merged**